SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp

```diff
@@ -0,0 +1,239 @@
+  static constexpr atomic_op ChosenOperator = ChosenOperatorT::value;
+  static constexpr functions::offset_generation AlgorithmToChange =
+      AlgorithmToChangeT::value;
+  using TestDescriptionT = functions::TestDescription<NumElems, TestCaseT>;
+
+public:
+  bool operator()(sycl::queue &queue, const std::string &data_type,
+                  const std::string &operator_type) {
+    static_assert(NDRangeDim <= 3 && NDRangeDim > 0);
+    static_assert(NumElemsToChange <= NumElems);
+    bool passed = true;
+    constexpr DataT base_value = 100;
+    constexpr DataT result_fill_value = 5;
+    shared_allocator<DataT> allocator(queue);
+    shared_vector<DataT> shared_init_values(NumElems, allocator);
+    functions::fill_init_values<NumElems, ChosenOperator, DataT>(
+        base_value, shared_init_values);
+
+    shared_vector<size_t> shared_offsets(NumElemsToChange,
+                                         shared_allocator<size_t>(queue));
+    functions::fill_offsets<NumElemsToChange, AlgorithmToChange>(
+        shared_offsets);
+    shared_vector<size_t> local_offsets = shared_offsets;
+    constexpr int NumberIterations = 16;
+    shared_vector<DataT> shared_result(NumberIterations, result_fill_value,
+                                       allocator);
+    sycl::nd_range<NDRangeDim> nd_range =
+        get_sycl_nd_range<NDRangeDim>(NumberIterations);
+
+    queue.submit([&](sycl::handler &cgh) {
+      DataT *shared_init_values_ptr = shared_init_values.data();
+      DataT *shared_result_ptr = shared_result.data();
+      size_t *const offsets_ptr = shared_offsets.data();
+
+      cgh.parallel_for<
+          Kernel<DataT, NumElems, TestCaseT, ChosenOperatorT, NDRangeDimT,
+                 NumElemsToChangeT, ChangeElemFilterT, AlgorithmToChangeT>>(
+          nd_range, [=](sycl::nd_item<1> nd_item) SYCL_ESIMD_KERNEL {
+            const size_t work_item_index = nd_item.get_global_linear_id();
+            shared_result_ptr[work_item_index] =
+                TestCaseT::template call_esimd_function<
+                    DataT, NumElemsToChange, ChangeElemFilterT, ChosenOperator>(
+                    shared_init_values_ptr, offsets_ptr);
+          });
+    });
+    queue.wait_and_throw();
+
+    ChangeElemFilterT filter{};
+    size_t num_changed_elems = 0;
+    // Collect the number of elements that should be changed. This value will be
+    // used to calculate expected value for all elements that will be returned
+    // after atomic_update call the sum of changed elements.
+    for (size_t i = 0; i < NumElemsToChange; ++i) {
+      num_changed_elems += (filter(i) == 1);
+    }
+
+    std::sort(shared_result.begin(), shared_result.end());
+    for (size_t i = 0; i < NumberIterations; ++i) {
+      const DataT &retrieved = shared_result[i];
+      const DataT &expected = base_value * num_changed_elems +
+                              (i - NumberIterations) * num_changed_elems;
```

Comment:
1. As a minor suggestion:

```suggestion
      const DataT &expected = (base_value - NumberIteractions + i) * num_changed_elems;
```
--------------------
2. So that's a sorted sum of values before update. Shouldn't this value be different for `increment` and `decrement` operators? You are starting from `shared_init_values` filled with `base_value` for both.

Reply:
the new algorithm to calculate expected value for this case that depends on `atomic_op` enum value has been implemented in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/b691d458173b14ad5b2997513cda4d914a364989